cmd/config/set.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/spf13/cast"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func newSetCommand(vp *viper.Viper) *cobra.Command {
+	return &cobra.Command{
+		Use:   "set KEY VALUE",
+		Short: "Sets an individual value in the hubble config file",
+		Long:  "Sets an individual value in the hubble config file",
+		RunE: func(_ *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				return fmt.Errorf("invalid arguments: set requires exactly 2 arguments: got '%s'", strings.Join(args, " "))
+			}
+			return runSet(vp, args[0], args[1])
+		},
+	}
+}
+
+func runSet(vp *viper.Viper, key, value string) error {
+	if !vp.InConfig(key) {
+		return fmt.Errorf("unknown key: %s", key)
+	}
+
+	// detect the type that we expect for the value
+	// note that only types supported by Viper need to be covered
+	var newVal interface{}
+	switch v := reflect.TypeOf(vp.Get(key)); v.Kind() {
+	case reflect.Bool:
+		b, err := cast.ToBoolE(value)
+		if err != nil {
+			return fmt.Errorf("cannot assign value for k=%s: expected type=bool, got type=%T", key, value)
+		}
+		newVal = b
+	case reflect.Float64:
+		f, err := cast.ToFloat64E(value)
+		if err != nil {
+			return fmt.Errorf("cannot assign value for k=%s: expected type=float64, got type=%T", key, value)
+		}
+		newVal = f
+	case reflect.Int:
+		i, err := cast.ToIntE(value)
+		if err != nil {
+			return fmt.Errorf("cannot assign value for k=%s: expected type=int, got type=%T", key, value)
+		}
+		newVal = i
+	case reflect.Slice:
+		switch iv := v.Elem(); iv.Kind() {
+		// assume we want a slice of string as we have no use for a slice of
+		// int anyways
+		case reflect.Interface:
+			var ss []string
+			for _, s := range strings.Split(value, ",") {
+				ss = append(ss, strings.TrimSpace(s))
+			}
+			newVal = ss
+		default:
+			return fmt.Errorf("unhandeld kind %s of %s, please open an issue", v.Kind(), iv.Kind())
+		}
+	case reflect.String:
+		newVal = value
+		// FIXME: find a way to figure out which keys are of type duration
```

Comment:
🤔

Reply:
Values of type `time.Duration` are stored as string (eg:  "3s"). If we don't do this check, a user can `hubble config set timeout foobar`, which works, but then the value is unusable. Not sure of a good way to deal with this atm.

Reply:
Yeah I understand the problem. I suppose there is no harm to display them as-is, right? If timeout is set to `5h` displaying that string shouldn't be confusing to the user.

Why do we have to cast necessarily?

Reply:
Deleted my previous comment, because I misunderstood the function.

Did you lift this function somewhere from cobra-viper integration, or did you write it from scratch? Somewhere in the cobra-viper path there should be something very similar on casting strings to their correct types.

Reply:
I wrote it from scratch. As we bind viper to cobra flags, I figured out a more robust way to assign the correct type for the value. As an added bonus, one can now omit providing a value which will reset it to its default. This new capability also prompted me to add `hubble config reset`.